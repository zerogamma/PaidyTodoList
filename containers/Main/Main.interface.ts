export interface todoType {
    id:number,
    text: string
}