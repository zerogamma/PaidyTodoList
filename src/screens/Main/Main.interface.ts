export interface todoType {
  id: number;
  text: string;
  status: boolean;
}
