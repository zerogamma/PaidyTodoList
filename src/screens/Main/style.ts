import { StyleSheet } from "react-native";
// Utils
import colors from "../../../assets/colors/color";

export const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  taskWrapper: {
    paddingTop: 120,
    paddingHorizontal: 40,
  },
  title: {
    fontSize: 24,
    fontWeight: "bold",
  },
  items: {},
  inputTaskWrapper: {
    position: "absolute",
    bottom: 30,
    width: "100%",
    paddingHorizontal: 35,
    flexDirection: "row",
    justifyContent: "space-around",
    alignItems: "center",
  },
  inputTask: {
    width: 250,
    backgroundColor: colors.secondary,
    borderRadius: 60,
    paddingVertical: 15,
    paddingHorizontal: 15,
    borderColor: colors.third,
    borderWidth: 1,
  },
  addText: {
    color: colors.third,
    fontSize: 60,
  },
});
