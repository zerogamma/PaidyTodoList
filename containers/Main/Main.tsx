import React from "react";
import {
  Keyboard,
  KeyboardAvoidingView,
  Platform,
  StyleSheet,
  Text,
  TextInput,
  TouchableOpacity,
  View,
} from "react-native";
// Components
import { LinearGradient } from "expo-linear-gradient";
import { Ionicons } from "@expo/vector-icons";
// Custom Components
import Task from "../../components/Task";
// Styles
import { styles } from "./style";
// Utils
import colors from "../../assets/colors/color";
import { useState } from "react";
import { todoType } from "./Main.interface";
import { useEffect } from "react";
import { useRef } from "react";

const UPDATE_INITIAL_STATE = {
  state: false,
  position: 0,
};

export default function Main() {
  const [task, setTask] = useState<string>("");
  const [todoList, setTodoList] = useState<todoType[]>([]);
  const [newItemId, setNewItemId] = useState<number>(0);
  const [update, setUpdate] = useState<{ state: boolean; position: number }>(
    UPDATE_INITIAL_STATE
  );
  const TEXT_REF = useRef<TextInput>(null);

  useEffect(() => {
    setNewItemId(todoList.length + 1);
  }, []);

  const handleAddTask = () => {
    Keyboard.dismiss();
    if (update.state) {
      let copyList = [...todoList];
      copyList[update.position].text = task;
      setTodoList(copyList);
      setTask("");
      setUpdate(UPDATE_INITIAL_STATE)
      return;
    }
    setTodoList([...todoList, { id: newItemId, text: task }]);
    setNewItemId(newItemId + 1);
    setTask("");
  };

  const handleRemoveTask = (index: number) => {
    const copyList = [...todoList];
    copyList.splice(index, 1);
    setTodoList(copyList);
  };

  const handleUpdateTask = (index: number, todo: string) => {
    setTask(todo);
    setUpdate({ state: true, position: index });
    TEXT_REF?.current && TEXT_REF.current.focus();
  };

  return (
    <View style={styles.container}>
      <LinearGradient
        colors={colors.primaryGradient}
        style={styles.background}
      />
      <View style={styles.taskWrapper}>
        <Text style={styles.title}>Todo List</Text>
        <View style={styles.items}>
          {todoList.map((todo, index) => (
            <Task
              key={todo.id}
              todo={todo.text}
              position={index}
              remove={handleRemoveTask}
              update={handleUpdateTask}
            />
          ))}
        </View>
      </View>

      <KeyboardAvoidingView
        behavior={Platform.OS === "ios" ? "padding" : "height"}
        style={styles.inputTaskWrapper}
      >
        <TextInput
          ref={TEXT_REF}
          style={styles.inputTask}
          placeholder={"Enter new task"}
          value={task}
          onChangeText={setTask}
        />
        <TouchableOpacity onPress={handleAddTask}>
          <Ionicons
            name={update.state ? "create" : "add-circle"}
            style={styles.addText}
          />
        </TouchableOpacity>
      </KeyboardAvoidingView>
    </View>
  );
}
