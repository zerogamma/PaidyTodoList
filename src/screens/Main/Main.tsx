import React, { useState, useEffect, useRef, useContext } from "react";
import {
  Keyboard,
  KeyboardAvoidingView,
  Platform,
  Text,
  TextInput,
  TouchableOpacity,
  View,
  FlatList,
  Animated,
  Dimensions,
} from "react-native";
// Context
import todoContext from "../../context/context";
// Components
import { Ionicons } from "@expo/vector-icons";
// Custom Components
import Task from "../../components/Task";
// Styles
import { styles } from "./style";
import { todoType } from "./Main.interface";
import { orderDesc } from "../../utils/utils";

// initial update value used to reset after the update is completed.
const UPDATE_INITIAL_STATE = {
  state: false,
  id: 0,
};

export function Main() {
  // Global state mainly data
  const apiTask = useContext(todoContext);
  // inner state management
  const [task, setTask] = useState<string>("");
  const [todoList, setTodoList] = useState<todoType[]>([]);
  const [newItemId, setNewItemId] = useState<number>(0);
  const [update, setUpdate] = useState<{ state: boolean; id: number }>(
    UPDATE_INITIAL_STATE
  );
  const [isKeyboardVisible, setKeyboardVisible] = useState<boolean>(false);

  // reference to get control specific component in this case the input to trigger is there and update
  const TEXT_REF = useRef<TextInput>(null);

  useEffect(() => {
    const keyboardDidShowListener = Keyboard.addListener(
      "keyboardDidShow",
      () => {
        setKeyboardVisible(true);
      }
    );
    const keyboardDidHideListener = Keyboard.addListener(
      "keyboardDidHide",
      () => {
        setKeyboardVisible(false);
      }
    );
    return () => {
      keyboardDidHideListener.remove();
      keyboardDidShowListener.remove();
    };
  }, []);

  // useEffect used to get ids for the new item. this if there is API no need to do it.
  // just using number to not make super complex the ID the ideal is have a better type or use an UUID
  useEffect(() => {
    if (todoList.length) {
      setNewItemId(todoList[0].id + 1);
    }
  }, [todoList]);

  // useEffect that check if there are any change on the data from the context
  // and added a sorting desc to always display the new item at the top.
  useEffect(() => {
    if (apiTask.todoList) {
      setTodoList(apiTask.todoList.sort(orderDesc));
    }
  }, [apiTask.todoList]);

  // sharring logic of the AddTask they can be separete but if I separate the component of the icon
  // but in this case for simplicity just used the same one and changed the icon look
  // thats why when the tigger of the update i made a state update to let know to the component is in update mode
  // every array manipulation is done in the context to make it close to what would happend if you have apollo/graph or
  // API where is more clean to have persistency layer.
  const handleAddTask = () => {
    // close the keyboard
    Keyboard.dismiss();
    // cancel if there is no text in the input.
    if(!task) return
    
    // the update function with a return to separate with the add logic.
    if (update.state) {
      apiTask.updateTask(task, update.id);
      setTask("");
      setUpdate(UPDATE_INITIAL_STATE);
      return;
    }
    apiTask.addNewTask({ id: newItemId, text: task, status: false });
    setTask("");
  };

  // remove
  const handleRemoveTask = (id: number) => {
    apiTask.deleteTask(id);
  };

  // check
  const handleStatusUpdate = (status: boolean, taskId: number) => {
    apiTask.updateTaskState(status,taskId);
  };

  // update
  const handleUpdateTask = (index: number, todo: string) => {
    setTask(todo);
    setUpdate({ state: true, id: index });
    TEXT_REF?.current && TEXT_REF.current.focus();
  };

  // code to give animation loading of the list.
  const translateX = useRef(
    new Animated.Value(Dimensions.get("window").height)
  ).current;

  useEffect(() => {
    Animated.timing(translateX, {
      toValue: 0,
      duration: 500,
      useNativeDriver: false,
    }).start();
  });

  // function to add animation view and used by FlatList to populate himself.
  const ItemView = ({ item, index }: { item: todoType; index: number }) => {
    return (
      <Animated.View style={{ transform: [{ translateY: translateX }] }}>
        <Task
          key={item.id}
          todo={item.text}
          id={item.id}
          checked={item.status}
          remove={handleRemoveTask}
          update={handleUpdateTask}
          updateCheck={handleStatusUpdate}
        />
      </Animated.View>
    );
  };

  return (
    <View style={styles.container}>
      <View style={styles.taskWrapper}>
        <Text style={styles.title}>Todo List</Text>
        <View
          style={[styles.items, { height: isKeyboardVisible ? "80%" : "88%" }]}
        >
          <FlatList
            data={todoList}
            renderItem={ItemView}
            keyExtractor={(item, index) => index.toString()}
          />
        </View>
      </View>

      <KeyboardAvoidingView
        behavior={Platform.OS === "ios" ? "padding" : "height"}
        style={styles.inputTaskWrapper}
      >
        <TextInput
          ref={TEXT_REF}
          style={styles.inputTask}
          placeholder={"Enter new task"}
          value={task}
          onChangeText={setTask}
        />
        <TouchableOpacity onPress={handleAddTask} testID="addTask">
          <View style={styles.iconWrapper}>
            <Ionicons
              name={update.state ? "create" : "add-circle"}
              style={styles.addText}
            />
          </View>
        </TouchableOpacity>
      </KeyboardAvoidingView>
    </View>
  );
}
