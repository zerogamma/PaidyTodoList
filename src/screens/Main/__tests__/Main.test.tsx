import React from "react";
import { Main } from "../Main";
import { render } from "../../../context/Main.test-utils";
import { fireEvent, cleanup, act } from "@testing-library/react-native";

const MainComponent = () => {
  return <Main />;
};

describe("<TaskComponent />", () => {
  afterEach(cleanup);

  it("renders correctly", () => {
    const container = render(<MainComponent />);
    expect(container).toMatchSnapshot();
  });
  it("Should not match the snashot for the input value", () => {
    const testText = "Testing input Value";
    const { getByPlaceholderText } = render(<MainComponent />);
    const inputComp = getByPlaceholderText("Enter new task");
    expect(inputComp.props.value).toEqual("");
    fireEvent.changeText(inputComp, testText);
    expect(inputComp.props.value).toEqual(testText);
  });
  it("Should add a new Task", () => {
    const { getByTestId, getByText, getByPlaceholderText } =
      render(<MainComponent />);
    const newTaskTest = "testing new Task";
    const inputComp = getByPlaceholderText("Enter new task");
    fireEvent.changeText(inputComp, newTaskTest);
    const addTask = getByTestId("addTask");
    fireEvent.press(addTask);
    const newTask = getByText(newTaskTest);
    expect(newTask).toBeTruthy();
  });
});
