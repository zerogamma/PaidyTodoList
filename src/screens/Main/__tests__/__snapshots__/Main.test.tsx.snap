// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskComponent /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "paddingHorizontal": 40,
        "paddingTop": 120,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
        }
      }
    >
      Todo List
    </Text>
    <View
      style={Object {}}
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "bottom": 30,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "paddingHorizontal": 35,
          "position": "absolute",
          "width": "100%",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Enter new task"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderColor": "#0068f2",
          "borderRadius": 60,
          "borderWidth": 1,
          "paddingHorizontal": 15,
          "paddingVertical": 15,
          "width": 250,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="addTask"
    >
      <Text />
    </View>
  </View>
</View>
`;
